Sources/Algorithms/Chunked.swift

```diff
@@ -246,3 +246,136 @@ extension Collection {
+///
+/// A ChunkedCollection is a lazy view on the base Collection, but it does not implicitly confer
+/// laziness on algorithms applied to its result.  In other words, for ordinary collections `c`:
+///
+/// * `c.chunks(of: 3)` does not create new storage
+/// * `c.chunks(of: 3).map(f)` maps eagerly and returns a new array
+/// * `c.lazy.chunks(of: 3).map(f)` maps lazily and returns a `LazyMapCollection`
+public struct ChunkedCollection<Base: Collection> {
+  
+  public typealias Element = Base.SubSequence
+  
+  public let base: Base
+  @usableFromInline
+  internal let _chunkCount: Int
+  
+  ///  Creates a view instance that presents the elements of `base`
+  ///  in `SubSequence` chunks of the given size.
+  ///
+  /// - Complexity: O(1)
+  @inlinable
+  internal init(_base: Base, _chunkCount: Int) {
+    self.base = _base
+    self._chunkCount = _chunkCount
+  }
+}
+
+extension ChunkedCollection: Collection {
+  public struct Index {
+    public let base: Base.Index
+    
+    @usableFromInline
+    init(_base: Base.Index) {
+      self.base = _base
+    }
+  }
+  
+  public var startIndex: Index { Index(_base: base.startIndex) }
+  public var endIndex: Index { Index(_base: base.endIndex) }
+  
+  public subscript(i: Index) -> Element {
+    base[i.base..<index(after: i).base]
+  }
+  
+  @inlinable
+  public func index(after i: Index) -> Index {
+    let baseIdx = base.index(
+      i.base, offsetBy: _chunkCount, limitedBy: base.endIndex)
+    return Index(_base: baseIdx ?? base.endIndex)
+  }
+}
+
+extension ChunkedCollection.Index: Comparable {
+  @inlinable
+  public static func < (lhs: ChunkedCollection.Index,
+                        rhs: ChunkedCollection.Index) -> Bool {
+    lhs.base < rhs.base
+  }
+}
+
+extension ChunkedCollection:
+  BidirectionalCollection, RandomAccessCollection
+where Base: RandomAccessCollection {
+  @inlinable
+  public func index(before i: Index) -> Index {
+    if i.base == base.endIndex {
+      let remainder = base.count%_chunkCount
+      if remainder != 0 {
+        return Index(_base: base.index(i.base, offsetBy: -remainder))
+      }
+    }
+    return Index(_base: base.index(i.base, offsetBy: -_chunkCount))
+  }
+  
+  @inlinable
+  public func distance(from start: Index, to end: Index) -> Int {
+    let distance = base.distance(from: start.base, to: end.base)
+    let (quotient, remainder) =
+      distance.quotientAndRemainder(dividingBy: _chunkCount)
+    return quotient + remainder
+  }
+  
+  @inlinable
+  public func index(_ i: Index, offsetBy n: Int) -> Index {
+    guard n != 0 else { return i }
+    
+    let bound = n > 0 ? base.endIndex : base.startIndex
+    return Index(_base:
+      base.index(i.base, offsetBy: n * _chunkCount, limitedBy: bound) ?? bound
+    )
+  }
+  
+  @inlinable
+  public var count: Int {
+    return base.count
+      .isMultiple(of: _chunkCount) ? base.count/_chunkCount
+                                   : base.count/_chunkCount + 1
```

Comment:
`count` and `distance` seem to be doing the same calculation, just expressed with different code. If we're going to repeat ourselves, we should at least use the same code, though it's probably best to try to avoid repeating ourselves

Reply:
Ah I just noted that default implementation of collection `count` is in terms of distance, so I think we don't even need this

Reply:
Since `count` is a customization point for `Collection`, it’s good to implement it in wrapper collections that can directly calculate the new count.

Reply:
Ah ok, so I'll add it back :)

Reply:
Added!